### ISSUE TITLE
DNS Resolution randomly fails in 3.6.0

### ISSUE THREAD

**Issue by Starfox64** [docker/for-win#12018]
- [X] I have tried with the latest version of Docker Desktop
  - [X] I have tried disabling enabled experimental features
  - [X] I have uploaded Diagnostics
  - Diagnostics ID: D713214F-FFC9-4294-A2F2-AD795FF255E4/20210827151944

### Actual behavior
DNS Resolution of some microsoft domain names (login.microsoftonline.com & graph.microsoft.com) randomly fails inside containers since updating to Docker For Windows 3.6.0.

This happens consistently for a few minutes after starting a container then works again intermittently.
### Expected behavior
Domain names resolve correctly.
### Information
I suspect this is coming from the fix mentioned in the release note:
`Fixed a bug where the DNS server would fail after receiving an unexpectedly large datagram.`

  - Windows Version: Windows 10 21H1
  - Docker Desktop Version: 3.6.0
  - WSL2 or Hyper-V backend?  WSL2
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: No

### Steps to reproduce the behavior

  1. `docker run -it --rm php:7.4-apache-buster bash`
  2. `curl https://graph.microsoft.com`
  3. Could no resolve host name

**Comment by dave-br** [comment]
I appear to be having the same problem on 3.6.0.  Kestrel / ASP.NET needs to access https://login.microsoftonline.com/Common/.well-known/openid-configuration, and fails to resolve (as does curl from inside the container, whereas other non-microsoft hosts appear to resolve fine).  Replacing the nameserver line in /etc/resolv.conf to 8.8.8.8 (from 127.0.0.11) helps, at least until the container is tossed and a new one is recreated at which point the default /etc/resolv.conf is reinstated.

I am NOT using WSL 2, in case that data point is helpful

**Comment by fadamsen** [comment]
We're having the same problem _just_ after updating to 3.6.0 today.

**Comment by fadamsen** [comment]
Docker Desktop for Windows 4.0.0 still has this issue. Downgrading to 3.5.2 fixed it for us, but it is obviously not tenable to be unable to upgrade. Is Docker Desktop for Windows a priority at all?

**Comment by DaleyKD** [comment]
I'm thankful to see this issue.  I had been on 3.6.0 for QUITE a while, and it just STOPPED today.  Out of nowhere.  I've been fighting it all day and even tried 4.0.0.  Going to try and rollback to 3.5.2 just so I'm not stuck.

Gotta say, this is absolutely nuts.  Trying to do `ping login.microsoftonline.com` from within a container doesn't work, but `ping cnn.com` does.

**Comment by DaleyKD** [comment]
Downgraded to 3.5.2 and I'm chugging along now.  Would love to see this fixed ASAP.

**Comment by kindanam** [comment]
Same here.
We're having the same issue while connecting to Azure Key Vault (so Microsoft obviously).
Downgrading to 3.5.2 for the moment.

**Comment by cowlinb6** [comment]
@kindanam as a workaround for accessing keyvault you can add DNS overrides into your hosts file. A bit of a hack but lets you continue to use the later versions of Docker.

20.190.160.4 login.windows.net
20.190.160.2 login.microsoftonline.com

**Comment by Starfox64** [comment]
@stephen-turner Have you guys been able to identify the cause? As said in the OP I suspect this is coming from the fix mentioned in the release notes for 3.6.0:
`Fixed a bug where the DNS server would fail after receiving an unexpectedly large datagram.`

After debugging some Microsoft DNS responses it seems they return a lot more RRs than is typical.

**Comment by jithurjacob** [comment]
Facing same issue with WSL2. Downgrading to 3.5.2 helped. 

Please fix this.

**Comment by SmartRPAJS** [comment]
I also started having this issue after upgrading to 3.6.0. 
Downgrading to 3.5.2 fixed it.

**Comment by DaleyKD** [comment]
I see they released 4.0.1. I didn't see this mentioned in the changelog.  Any updates?

**Comment by hafkenscheid** [comment]
Nope, still issues with 4.0.0 and 4.0.1

**Comment by stephen-turner** [comment]
We're trying to look into this, but we can't reproduce it. E.g.:
```
$ docker run alpine ping login.microsoftonline.com
PING login.microsoftonline.com (40.126.31.143): 56 data bytes
```
Can someone give me a single-line command that reliably fails?

**Comment by dave-br** [comment]
In my case, it's just a single machine sitting in my home.  Nothing needs to be deployed to the cloud.  All I need to run is 

curl https://login.microsoftonline.com

But the thing is that this is intermittent.  It happens very early after the container has been created, and stays bad for a while (e.g., 20 minutes).  And then it works for a while, and then it goes bad again.

I've just tried this again from my downgraded container (v. 3.5.2) where everything works fine, and I see that login.microsoftonline.com redirects to https://www.office.com/login#, and the latter returns no html via curl.  This seems new to me, and I THOUGHT curl returned a lot of html directly from https://login.microsoftonline.com, back when I had the newer, failing version of Docker.  But I don't really trust my memory, so this behavior might not really be any different.

But it does highlight that it's problematic tying a reproducing scenario to a web site / DNS server you don't control.  Starfox64's theory might be something you can pursue--if the DNS is not reliably returning as large a datagram as it may have before, can you hack a scenario where you force docker to think it is, and try to use that to investigate somehow?

**Comment by SmartRPAJS** [comment]
@stephen-turner 
> We're trying to look into this, but we can't reproduce it. E.g.:
> 
> ```
> $ docker run alpine ping login.microsoftonline.com
> PING login.microsoftonline.com (40.126.31.143): 56 data bytes
> ```
> 
> Can someone give me a single-line command that reliably fails?

For me the issue only appears when using docker-compose. I can replicate it across multiple desktops this way:

1. Create a `docker-compose.yml` file with the contents:

```yml
version: '3'
services:
    alpine:
        image: alpine
        command: tail -F anything
```

2. Open a command prompt and run: `docker-compose up -d`
3. When the container is running use the `CLI` button on it in Docker Dashboard
4. Run the command `ping google.com -c 1`.
5. Observe it works.
6. Run the command `ping login.microsoftonline.com -c 1`.
7. It fails with the message: `ping: bad address 'login.microsoftonline.com'`

**Comment by stephen-turner** [comment]
I followed this recipe, but unfortunately(?) step 6 succeeds for me:

```
/ # ping login.microsoftonline.com -c 1
PING login.microsoftonline.com (40.126.31.137): 56 data bytes

--- login.microsoftonline.com ping statistics ---
1 packets transmitted, 0 packets received, 100% packet loss
```

**Comment by jithurjacob** [comment]
This could be tough to reproduce. I was getting this issue while running container from a DevTest Lab Windows 10 VM. A python script for accessing a Key Vault value was succeeding when run locally but was failing within a ubuntu Docker Container. 

I had some luck by adding the Corporate CA certificates in the container. Then the connections was working for me but still some times it failed. I am new to the organization so I am not aware of the exact network/firewall configurations. However I can help you in testing any specific settings.

Downgrading to 3.5.2 worked perfectly fine for me and didn't need to add any certificate or make any changes.

**Comment by DaleyKD** [comment]
I wish I had time to help you with possible reproduction steps, but I don't have time right now to battle upgrades, get something that causes the issue, then downgrade. For me, that'd take all day.

What @SmartRPAJS showed is pretty much what I would've done anyway.

Yes, I'm using docker compose for everything.

**Comment by SmartRPAJS** [comment]
@stephen-turner 
> I followed this recipe, but unfortunately(?) step 6 succeeds for me:
> 
> ```
> / # ping login.microsoftonline.com -c 1
> PING login.microsoftonline.com (40.126.31.137): 56 data bytes
> 
> --- login.microsoftonline.com ping statistics ---
> 1 packets transmitted, 0 packets received, 100% packet loss
> ```

Oh, that's too bad. Thank you for trying though.

I just tried to compare results of `nslookup` for the two ways of starting a container with the alpine image:

### Run

1. Start the container with the command: `docker run -it alpine` 
2. When the container is running use the `CLI` button on it in Docker Dashboard
3. Execute the command `nslookup -debug login.microsoftonline.com`
4. Result:

```
/ # nslookup -debug login.microsoftonline.com
Server:         192.168.65.5
Address:        192.168.65.5:53

Query #0 completed in 16ms:
Non-authoritative answer:
*** Can't find login.microsoftonline.com: Parse error

Query #1 completed in 102ms:
Non-authoritative answer:
login.microsoftonline.com       canonical name = ak.privatelink.msidentity.com
ak.privatelink.msidentity.com   canonical name = www.tm.ak.prd.aadg.akadns.net
```

### Compose

1. Use the `docker-compose.yml` (posted earlier).
2. When the container is running use the `CLI` button on it in Docker Dashboard
3. Execute the command `nslookup -debug login.microsoftonline.com`
4. Result:

```
/ # nslookup -debug login.microsoftonline.com
Server:         127.0.0.11
Address:        127.0.0.11:53

Query #0 completed in 42ms:
Non-authoritative answer:
*** Can't find login.microsoftonline.com: No answer

Query #1 completed in 212ms:
Non-authoritative answer:
login.microsoftonline.com       canonical name = ak.privatelink.msidentity.com
ak.privatelink.msidentity.com   canonical name = www.tm.ak.prd.aadg.trafficmanager.net
ak.privatelink.msidentity.com   canonical name = www.tm.ak.prd.aadg.akadns.net
```

Notice the difference in Query #0 for the two containers:

* Can't find login.microsoftonline.com: Parse error
* Can't find login.microsoftonline.com: No answer

**Comment by stephen-turner** [comment]
Interestingly I get something completely different:

```
/ # nslookup -debug login.microsoft.com
Server:         127.0.0.11
Address:        127.0.0.11:53

Query #0 completed in 26ms:
Non-authoritative answer:
login.microsoft.com     canonical name = a.privatelink.msidentity.com
a.privatelink.msidentity.com    canonical name = prda.aadg.msidentity.com
prda.aadg.msidentity.com        canonical name = www.tm.a.prd.aadg.trafficmanager.net
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.6
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.141
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 20.190.159.132
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.137
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 20.190.159.134
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.1
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 20.190.159.136
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.143
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.141
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 20.190.159.132
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.137
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 20.190.159.134
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.1
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 20.190.159.136
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.143
Name:   www.tm.a.prd.aadg.trafficmanager.net
Address: 40.126.31.6

Query #1 completed in 178ms:
Non-authoritative answer:
login.microsoft.com     canonical name = a.privatelink.msidentity.com
a.privatelink.msidentity.com    canonical name = prda.aadg.msidentity.com
prda.aadg.msidentity.com        canonical name = www.tm.a.prd.aadg.trafficmanager.net
```

I'm beginning to wonder if this bug only manifests in some parts of the world because the DNS response is geo-dependent. (I'm in the UK).

**Comment by DaleyKD** [comment]
> I'm beginning to wonder if this bug only manifests in some parts of the world because the DNS response is geo-dependent. (I'm in the UK).

@stephen-turner : That may very well be. I'm in the Dallas, TX area, and a teammate of mine who had the same issue was in Cleveland, OH area. I also have teammates in the Houston, TX and St. Louis, MO area that haven't had this issue.

**Comment by SmartRPAJS** [comment]
> I'm beginning to wonder if this bug only manifests in some parts of the world because the DNS response is geo-dependent. 

@stephen-turner Do you have any idea, why I see the difference between `compose` and `run` on the same machine then (I'm in Denmark)?

**Comment by SmartRPAJS** [comment]
I have created a better example now where I can show the difference using only `docker-compose`.

1. Create a `docker-compose.yml` file with the contents:
```
services:
    alpine_host:
        image: alpine
        command: tail -F anything
        network_mode: host
    alpine_network:
        image: alpine
        command: tail -F anything
```
2. Open a command prompt and run: `docker-compose up -d`
3. This will create two containers with same image, but one of them in `host` network mode.
4. When the containers are running, use the `CLI` button on both in Docker Dashboard.
5. In each terminal, execute the command `ping login.microsoftonline.com`
    1. Container `alpine_network` fails with the mesage: `ping: bad address 'login.microsoftonline.com'`
    2. Container `alpine_host` works with the mesage: `PING login.microsoftonline.com (40.126.31.140): 56 data bytes`
6. In each terminal, execute the command `cat /etc/resolv.conf`
    1. Container `alpine_network` contains:
        ```
        nameserver 127.0.0.11
        options ndots:0
        ```
    2. Container `alpine_host` contains:
        ```
        # DNS requests are forwarded to the host. DHCP DNS options are ignored.
        nameserver 192.168.65.5
        ```

**Comment by Drol** [comment]
Same problem here from Sweden.

Running this from ASP.NET inside docker never fails for google.com but sometimes randomly fails for login.microsoftonline.com.
```
var googleEntry = System.Net.Dns.GetHostEntry("google.com");
var loginEntry = System.Net.Dns.GetHostEntry("login.microsoftonline.com");
```

nslookup from WSL2 Ubuntu.
```
nslookup login.microsoftonline.com
Server:         1.1.1.1
Address:        1.1.1.1#53

Non-authoritative answer:
login.microsoftonline.com       canonical name = ak.privatelink.msidentity.com.
ak.privatelink.msidentity.com   canonical name = www.tm.ak.prd.aadg.trafficmanager.net.
Name:   www.tm.ak.prd.aadg.trafficmanager.net
Address: 20.190.159.133
Name:   www.tm.ak.prd.aadg.trafficmanager.net
Address: 40.126.31.136
Name:   www.tm.ak.prd.aadg.trafficmanager.net
Address: 40.126.31.142
Name:   www.tm.ak.prd.aadg.trafficmanager.net
Address: 40.126.31.5
Name:   www.tm.ak.prd.aadg.trafficmanager.net
Address: 40.126.31.9
Name:   www.tm.ak.prd.aadg.trafficmanager.net
Address: 40.126.31.138
Name:   www.tm.ak.prd.aadg.trafficmanager.net
Address: 40.126.31.2
Name:   www.tm.ak.prd.aadg.trafficmanager.net
Address: 20.190.159.137
www.tm.ak.prd.aadg.trafficmanager.net   canonical name = dub1.a.ak.prd.aadg.trafficmanager.net.
```

**Comment by stephen-turner** [comment]
@anyname2 Thanks, that looks like a good clue.

Yes, I'm on WSL 2. But I don't seem to be able to reach that DNS server.

Just to be sure, the DNS lookup succeeds from the WSL 2 command line, does it? I assume it does, but I don't think anyone's said so explicitly yet.

**Comment by stephen-turner** [comment]
@anyname2 In that case, I can't reproduce it, that's the same server address I have.

Also, are you on the latest Docker Desktop?

Are other people with this bug finding that it succeeds in ubuntu but fails in alpine? musl has a different DNS implementation that we've had problems with in the past.

**Comment by DaleyKD** [comment]
We just changed ISPs from Spectrum to AT&T Fiber. With IPv6 on, still same issues.  With IPv6 off, I'm fixed.  I can't remember if I ensured that IPv6 was off when I was with Spectrum or not.

Can you investigate an IPv6 cause?

**Comment by DaanBaars** [comment]
Same issue here all versions after 3.5.2, quick fix by disable IPV6 works

**Comment by p1-0tr** [comment]
Hi,

We believe the issue should have been fixed in the most recent (experimental) build of Docker Desktop, which you can get here - [DesktopInstaller](https://desktop-stage.docker.com/win/main/amd64/71379/Docker%20Desktop%20Installer.exe) . I would greatly appreciate if you could give it a go, and let me know if the issue has been resolved for you.

Thanks, Piotr

**Comment by DaanBaars** [comment]
@p1-0tr i can confirm that running that version with ipv6 enabled solves the issue

**Comment by p1-0tr** [comment]
@DaanBaars Thanks for your feedback :)

**Comment by fadamsen** [comment]
> We believe the issue should have been fixed in the most recent (experimental) build of Docker Desktop, which you can get here - [DesktopInstaller](https://desktop-stage.docker.com/win/main/amd64/71379/Docker%20Desktop%20Installer.exe) . I would greatly appreciate if you could give it a go, and let me know if the issue has been resolved for you.

Hi @p1-0tr, the linked build also works for us! 👍

**Comment by Drol** [comment]
> Hi,
> 
> We believe the issue should have been fixed in the most recent (experimental) build of Docker Desktop, which you can get here - [DesktopInstaller](https://desktop-stage.docker.com/win/main/amd64/71379/Docker%20Desktop%20Installer.exe) . I would greatly appreciate if you could give it a go, and let me know if the issue has been resolved for you.
> 
> Thanks, Piotr

Seems to work fine here too!

**Comment by kindanam** [comment]
Hi, any info about the availability of the fix ?

**Comment by Drol** [comment]
> Hi, any info about the availability of the fix ? Edit : I mean, in an official release ?

I see nothing in changelogs but latest version works without issues for me.

**Comment by p1-0tr** [comment]
Hi, the fix is available as of 4.3.0.
